lib/store/src/StoryIndexStore.ts

```diff
@@ -0,0 +1,87 @@
+import { Channel } from '@storybook/addons';
+import { StoryId } from '@storybook/csf';
+
+import { StorySpecifier, StoryIndex, StoryIndexEntry } from './types';
+
+type MaybePromise<T> = Promise<T> | T;
+
+export class StoryIndexStore {
+  fetchStoryIndex: () => MaybePromise<StoryIndex>;
+
+  channel: Channel;
+
+  stories: Record<StoryId, StoryIndexEntry>;
+
+  constructor({ fetchStoryIndex }: { fetchStoryIndex: StoryIndexStore['fetchStoryIndex'] }) {
+    this.fetchStoryIndex = fetchStoryIndex;
+  }
+
+  initialize(options: { sync: false }): Promise<void>;
+
+  initialize(options: { sync: true }): void;
+
+  initialize({ sync = false } = {}) {
+    return sync ? this.cache(true) : this.cache(false);
```

Comment:
why do we need this level of indirection between `initialize` and `cache`? and should we call it `_cache` if it's an internal method?

Reply:
I think we are quite likely to add some event handlers on a websocket in initialize down the track.